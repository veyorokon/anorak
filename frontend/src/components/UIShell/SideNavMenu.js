/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import ChevronDown20 from '@carbon/icons-react/lib/chevron--down/20';
import { settings } from 'carbon-components';
import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import SideNavIcon from './SideNavIcon';

const { prefix } = settings;

export class SideNavMenu extends React.Component {
  static propTypes = {
    /**
     * Provide an optional class to be applied to the containing node
     */
    className: PropTypes.string,

    /**
     * Provide <SideNavMenuItem>'s inside of the `SideNavMenu`
     */
    children: PropTypes.node,

    /**
     * Pass in a custom icon to render next to the `SideNavMenu` title
     */
    icon: PropTypes.node.isRequired,

    /**
     * Specify whether the `SideNavMenu` is "active". `SideNavMenu` should be
     * considered active if one of its menu items are a link for the current
     * page.
     */
    isActive: PropTypes.bool,

    /**
     * Provide the text for the overall menu name
     */
    title: PropTypes.string.isRequired,

    /**
     * Specify whether the menu should default to expanded. By default, it will
     * be closed.
     */
    defaultExpanded: PropTypes.bool,
  };

  static defaultProps = {
    defaultExpanded: false,
    isActive: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      isExpanded: props.defaultExpanded || false,
    };
  }

  handleToggleExpand = () => {
    this.setState(state => ({ isExpanded: !state.isExpanded }));
  };

  render() {
    const {
      buttonRef,
      className: customClassName,
      children,
      icon,
      isActive,
      title,
    } = this.props;
    const { isExpanded } = this.state;
    const className = cx({
      [`${prefix}--side-nav__item`]: true,
      [`${prefix}--side-nav__item--active`]: isActive,
      [customClassName]: !!customClassName,
    });
    return (
      <li className={className}>
        <button
          aria-haspopup="true"
          aria-expanded={isExpanded}
          className={`${prefix}--side-nav__submenu`}
          onClick={this.handleToggleExpand}
          ref={buttonRef}
          type="button">
          <SideNavIcon>{icon}</SideNavIcon>
          <span className={`${prefix}--side-nav__submenu-title`}>{title}</span>
          <SideNavIcon className={`${prefix}--side-nav__submenu-chevron`} small>
            <ChevronDown20 />
          </SideNavIcon>
        </button>
        <ul className={`${prefix}--side-nav__menu`} role="menu">
          {children}
        </ul>
      </li>
    );
  }
}

export default React.forwardRef((props, ref) => {
  return <SideNavMenu {...props} buttonRef={ref} />;
});
