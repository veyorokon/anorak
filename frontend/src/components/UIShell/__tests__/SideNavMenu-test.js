/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import { SideNavMenu } from '../SideNavMenu';

describe('SideNavMenu', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      buttonRef: jest.fn(),
      className: 'custom-classname',
      children: 'text',
      icon: <div>icon</div>,
      isActive: false,
      title: 'title',
    };
  });

  it('should render', () => {
    const wrapper = mount(<SideNavMenu {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should expand the menu when the button ref is clicked', () => {
    const wrapper = mount(<SideNavMenu {...mockProps} />);
    expect(wrapper.state('isExpanded')).toBe(false);
    expect(mockProps.buttonRef).toHaveBeenCalledTimes(1);
    wrapper.find('button').simulate('click');
    expect(wrapper.state('isExpanded')).toBe(true);
  });
});
