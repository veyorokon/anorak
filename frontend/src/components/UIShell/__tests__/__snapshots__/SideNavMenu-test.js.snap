// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavMenu should render 1`] = `
<SideNavMenu
  buttonRef={
    [MockFunction] {
      "calls": Array [
        Array [
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="bx--side-nav__submenu"
            type="button"
          >
            <div
              class="bx--side-nav__icon"
            >
              <div>
                icon
              </div>
            </div>
            <span
              class="bx--side-nav__submenu-title"
            >
              title
            </span>
            <div
              class="bx--side-nav__icon bx--side-nav__icon--small bx--side-nav__submenu-chevron"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                height="20"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16 22L6 12l1.4-1.4 8.6 8.6 8.6-8.6L26 12z"
                />
              </svg>
            </div>
          </button>,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  className="custom-classname"
  defaultExpanded={false}
  icon={
    <div>
      icon
    </div>
  }
  isActive={false}
  title="title"
>
  <li
    className="bx--side-nav__item custom-classname"
  >
    <button
      aria-expanded={false}
      aria-haspopup="true"
      className="bx--side-nav__submenu"
      onClick={[Function]}
      type="button"
    >
      <SideNavIcon
        small={false}
      >
        <div
          className="bx--side-nav__icon"
        >
          <div>
            icon
          </div>
        </div>
      </SideNavIcon>
      <span
        className="bx--side-nav__submenu-title"
      >
        title
      </span>
      <SideNavIcon
        className="bx--side-nav__submenu-chevron"
        small={true}
      >
        <div
          className="bx--side-nav__icon bx--side-nav__icon--small bx--side-nav__submenu-chevron"
        >
          <ChevronDown20
            height={20}
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 32 32"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <svg
              aria-hidden={true}
              focusable="false"
              height={20}
              preserveAspectRatio="xMidYMid meet"
              style={
                Object {
                  "willChange": "transform",
                }
              }
              viewBox="0 0 32 32"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 22L6 12l1.4-1.4 8.6 8.6 8.6-8.6L26 12z"
              />
            </svg>
          </ChevronDown20>
        </div>
      </SideNavIcon>
    </button>
    <ul
      className="bx--side-nav__menu"
      role="menu"
    >
      text
    </ul>
  </li>
</SideNavMenu>
`;
